File: leaf-android/Leaf.java
package leaf;

public class Leaf {
    public interface Callback {
        void reportTraffic(int txRate, int rxRate, long txTotal, long rxTotal);
        void reportState(int state);
    }

    public static native int runWithOptions(
            int rtId,
            final String configPath,
            Callback callback,
            boolean autoReload,
            boolean multiThread,
            boolean autoThreads,
            int threads,
            int stackSize);

    public static native int run(int rtId, final String configPath, Callback callback);

    public static native boolean reload(int rtId);

    public static native boolean shutdown(int rtId);

    public static native boolean testConfig(final String configPath);

    static {
        System.loadLibrary("leaf");
    }
}

File: leaf-darwin/leaf.swift
import Foundation
import LeafLib

public func runWithOptions(
    rtId: UInt16,
    configPath: String,
    callback: UnsafePointer<Callback>!,
    autoReload: Bool,
    multiThread: Bool,
    autoTheads: Bool,
    threads: Int32,
    stackSize: Int32
) -> Int32 {
    return leaf_run_with_options(rtId, configPath.cString(using: .utf8), callback, autoReload, multiThread, autoTheads, threads, stackSize);
}

public func run(rtId: UInt16, configPath: String, callback: UnsafePointer<Callback>!) -> Int32 {
    return leaf_run(rtId, configPath.cString(using: .utf8), callback)
}

public func reload(rtId: UInt16) -> Int32 {
    return leaf_reload(rtId)
}

public func shutdown(rtId: UInt16) -> Bool {
    return leaf_shutdown(rtId)
}
